stl/src/atomic_wait.cpp

```diff
@@ -0,0 +1,249 @@
+        return result;
+    }
+
+
+#if _STL_WIN32_WINNT >= _WIN32_WINNT_WIN8
+    constexpr bool _Have_wait_functions() {
+        return true;
+    }
+#define __crtWaitOnAddress       WaitOnAddress
+#define __crtWakeByAddressSingle WakeByAddressSingle
+#define __crtWakeByAddressAll    WakeByAddressAll
+
+#pragma comment(lib, "Synchronization.lib")
+
+    [[noreturn]] void _Atomic_wait_fallback(
+        [[maybe_unused]] const void* const _Storage, [[maybe_unused]] unsigned long long& _Wait_context) noexcept {
+        std::terminate();
+    }
+
+    [[noreturn]] void _Atomic_notify_fallback([[maybe_unused]] const void* const _Storage) noexcept {
+        std::terminate();
+    }
+
+    void _Atomic_unwait_fallback(
+        [[maybe_unused]] const void* const _Storage, [[maybe_unused]] unsigned long long& _Wait_context) noexcept {}
+
+#else // ^^^ _STL_WIN32_WINNT >= _WIN32_WINNT_WIN8 / _STL_WIN32_WINNT < _WIN32_WINNT_WIN8 vvv
+    void _Atomic_wait_fallback(const void* const _Storage, unsigned long long& _Wait_context) noexcept {
+        switch (_Wait_context & _Atomic_wait_phase_mask) {
+        case _Atomic_wait_phase_wait_none: {
+            _Wait_context = _Atomic_wait_phase_wait_locked;
+            auto& _Entry  = _Atomic_wait_table_entry(_Storage);
+            ::AcquireSRWLockExclusive(&_Entry._Lock);
+            // re-check, and still in _Atomic_wait_phase_wait_locked
+            break;
+        }
+
+        case _Atomic_wait_phase_wait_locked: {
+            auto& _Entry = _Atomic_wait_table_entry(_Storage);
+            ::SleepConditionVariableSRW(&_Entry._Condition, &_Entry._Lock, INFINITE, 0);
+            // re-check, and still in _Atomic_wait_phase_wait_locked
+            break;
+        }
+        }
+    }
+
+    void _Atomic_unwait_fallback(const void* const _Storage, unsigned long long& _Wait_context) noexcept {
+        if (_Wait_context & _Atomic_wait_phase_wait_locked) {
+            auto& _Entry = _Atomic_wait_table_entry(_Storage);
+            ::ReleaseSRWLockExclusive(&_Entry._Lock);
+        }
+    }
+
+    void _Atomic_notify_fallback(const void* const _Storage) noexcept {
+        auto& _Entry = _Atomic_wait_table_entry(_Storage);
+        ::AcquireSRWLockExclusive(&_Entry._Lock);
+        ::ReleaseSRWLockExclusive(&_Entry._Lock);
+        ::WakeAllConditionVariable(&_Entry._Condition);
+    }
+
+    struct _Wait_on_address_functions {
+        std::atomic<decltype(&::WaitOnAddress)> _Pfn_WaitOnAddress;
+        std::atomic<decltype(&::WakeByAddressSingle)> _Pfn_WakeByAddressSingle;
+        std::atomic<decltype(&::WakeByAddressAll)> _Pfn_WakeByAddressAll;
+        std::atomic<bool> _Initialized;
+    };
+
+    const _Wait_on_address_functions& _Get_wait_functions() {
+        static _Wait_on_address_functions functions;
+        if (!functions._Initialized.load(std::memory_order_acquire)) {
+            HMODULE sync_api_module        = ::GetModuleHandle(TEXT("API-MS-WIN-CORE-SYNCH-L1-2-0.DLL"));

```

Comment:
Should the DLL name be written in lowercase? Does it matter?

Reply:
I don't know. It is not officially documented.
[WaitOnAddress documentation ](https://docs.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-waitonaddress) lies about `Kernel32.dll`

Reply:
I mean, the DLL name was obtained by using Dependency Walker on statically linked `Synchronization.lib`. Names are always uppercase in this case

Reply:
`parallel_algorithms.cpp` uses `KernelBase.dll` for the same APIs

Reply:
It should be lowercase:
```
d:\Temp>type wait.cpp
#include <windows.h>

#pragma comment(lib, "Synchronization.lib")

int main()
{
    volatile int value = 1;
    int other_value = 1;
    ::WaitOnAddress(&value, &other_value, sizeof(int), INFINITE);
    return 0;
}
d:\Temp>cl wait.cpp
Microsoft (R) C/C++ Optimizing Compiler Version 19.26.28720.3 for x86
Copyright (C) Microsoft Corporation.  All rights reserved.

wait.cpp
Microsoft (R) Incremental Linker Version 14.26.28720.3
Copyright (C) Microsoft Corporation.  All rights reserved.

/out:wait.exe
wait.obj

d:\Temp>dumpbin /IMPORTS wait.exe|find /I "API-MS-WIN-CORE"
    api-ms-win-core-synch-l1-2-0.dll
```

Reply:
lowercased